Unified Storage: Remove wildcard queries

owensmallwood opened the pull request:
fixes https://github.com/grafana/search-and-storage-team/issues/218

**Context**
Were using Bleve wildcard queries in order to match substrings. With a large index, these queries are really slow and were causing pods to get OOMkilled. When testing with a 450k doc index, with a 30GB pod memory limit, it was reached after doing 5 or 6 consecutive queries 😬 .

If we want to match substrings, we should use an edge n-gram token filter. This is confirmed [here](https://groups.google.com/g/bleve/c/OsTZuCMnwUA). An edge n-gram filter creates a token for each consecutive substring anchored at one side of the token (the front in our case). Example: "hello" -> "hel", "hell", "hello". This way, if we want to match substrings, we'll be utilizing the index instead of brute-forcing it with a wildcard query.

**Important Changes**
- Introduced a new internal mapping for the title, `title_ngram`. This uses an edge ngram token filter so we don't have to do wildcard queries to match substrings.
- A search text query is now a disjunction of queries boosted in order of priority (exact match -> match phrase -> match). This helps to give us more relevant search results.
- When searching using a text query, results are sorted by score if no sort is present.
- When searching with no text query, results are sorted by title ascending (a -> b -> c).
- Adds lots of unit tests for title search as those were lacking.

**Functional Testing**
With these changes and the same 450k doc index, the OOM issues are gone locally.

owensmallwood committed:
WIP adding custom analyzer so we can do substring search efficiently

scottlepp commented on pkg/storage/unified/search/bleve.go:
```diff
@@ -561,7 +564,7 @@ func (b *bleveIndex) toBleveSearchRequest(ctx context.Context, req *resource.Res
 		Size:    int(req.Limit),
 		From:    int(req.Offset),
-		Explain: req.Explain,
+		Explain: true,
```
revert this?

scottlepp reviewed (commented):
I see this is draft so i'm guessing you'll be cleaning these items up anyway

owensmallwood commented:
@scottlepp yeah this is still a big WIP. I want to add a bunch more unit tests for search as well. Want to be use using this analyzer doesn't break stuff. Also, need to decide what other fields we want to use this analyzer for as well.

ryantxu commented on pkg/storage/unified/search/bleve_mappings.go:
```diff
@@ -1,19 +1,55 @@
 }
 
+func addNgramCustomAnalyzer(mapper *mapping.IndexMappingImpl) {
+	// Define an N-Gram tokenizer (for substring search)
```
Check: Maybe worth checking https://github.com/grafana/grafana/blob/main/pkg/services/searchV2/ngram.go#L30

for that we did 0-7 with ngrams, and then change the query based on length:
https://github.com/grafana/grafana/blob/main/pkg/services/searchV2/bluge.go#L464

owensmallwood committed:
Adding unit tests for title search

owensmallwood committed:
formatting

owensmallwood committed:
adds more title search unit tests

owensmallwood committed:
organize helpers

owensmallwood committed:
fixes issue caused by having two title mappings

owensmallwood committed:
Removes camelcase token filter since it prevents you from searching for a substring of chars and numbers. Adds regression test.

owensmallwood committed:
Merge branch 'main' into unified-storage-indexer-add-ngram-token-filter

owensmallwood committed:
adds back mapping for title_phrase

owensmallwood committed:
use simple analyzer for input query string so it doesn't filter out english stop words

owensmallwood committed:
ran bleve tests, table snapshots updated

owensmallwood committed:
ignore linter for "unused" test functions. They are very helpful for troubleshooting search. Keeping them.

owensmallwood committed:
only log total hits and query cost if result not nil

owensmallwood committed:
fixes failing test - one more field because there are two title mappings now

owensmallwood committed:
Merge branch 'main' into unified-storage-indexer-add-ngram-token-filter

owensmallwood committed:
fix test

owensmallwood committed:
fixes test - only take first item when its the title

owensmallwood committed:
Adds separate internal field for title ngram mapping.

When searching with a query, results are sorted by score desc.

When searching without a query, results are sorted by title desc.

Adjusts ngram max to be 10.

Text queries are a disjunction of an exact match, phrase match, and a match. Boosted to have priority in that order.

Adds more unit tests for searching.

owensmallwood committed:
linter

owensmallwood committed:
fix test

owensmallwood committed:
ran tests - generated new test dash json

owensmallwood committed:
sort by title phrase instead of title

owensmallwood committed:
fix test - not relying on /apis/dashboard/search to apply title sorting anymore

owensmallwood committed:
Merge branch 'main' into unified-storage-indexer-add-ngram-token-filter

owensmallwood commented on pkg/registry/apis/dashboard/search.go:
```diff
@@ -378,8 +382,8 @@ func (s *SearchHandler) DoSearch(w http.ResponseWriter, r *http.Request) {
 		// default sort by resource descending ( folders then dashboards ) then title
 		sort.Slice(parsedResults.Hits, func(i, j int) bool {
-			return parsedResults.Hits[i].Resource > parsedResults.Hits[j].Resource ||
-				(parsedResults.Hits[i].Resource == parsedResults.Hits[j].Resource && strings.ToLower(parsedResults.Hits[i].Title) < strings.ToLower(parsedResults.Hits[j].Title))
```
We should rely on Bleve to do sorting where we can. So this logic has been moved into the bleve.

owensmallwood commented on pkg/storage/unified/search/bleve.go:
```diff
@@ -596,12 +601,26 @@ func (b *bleveIndex) toBleveSearchRequest(ctx context.Context, req *resource.Res
+
+		// Query 1: Match the exact query string
+		queryExact := bleve.NewMatchQuery(req.Query)
+		queryExact.SetBoost(10.0)
```
I'm not sure what value to boost by. Seems arbitrary 🤷‍♂️

JohnnyQQQQ commented:
/deploy-to-hg
